content/posts/vault-migration/index.md

```diff
@@ -0,0 +1,558 @@
+
+To enable unseal key migration, we first need to configure the new Vault cluster
+to support dual seal configuration — keeping the old seal block active but
+disabled, and adding a new one that points to the new Azure Key Vault.
+
+- Update the Vault configuration:
+
+```bash
+kubectl get cm vault-config -o yaml > vault-config.yaml
+
+ed '+/seal "/' vault-config.yaml <<EOF
+a
+      disabled       = true
+.
+/}/
+a
+    seal "azurekeyvault" {
+      tenant_id      = "${TENANT_ID}"
+      client_id      = "${NEW_VAULT_MSI_CLIENT_ID}"
+      vault_name     = "${NEW_KEY_VAULT_NAME}"
+      key_name       = "${NEW_VAULT_UNSEAL_KEY_SECRET_NAME}"
+    }
+.
+w
+q
+EOF
+
+kubectl apply -f vault-config.yaml
+```
+
+### 8. Perform the unseal key migration on the follower pods
+
+Vault only allows unseal key migration on follower nodes, so we restart each
+follower pod and run the migration steps.
+
+- Iterate through each follower pod and migrate using all three unseal keys:
+
+  ```bash
+  for i in {0..2}; do
+    if [ "${i}" -eq "${leader_index}" ]; then
+      continue
+    fi
+
+    print_message "Restarting pod vault-${i} to start unseal key migration"
+    kubectl delete pod "vault-${i}"
+    kubectl wait \
+      --for=jsonpath='{.status.containerStatuses[0].state.running}' \
+      --selector=apps.kubernetes.io/pod-index="${i}" \
+      --timeout=90s \
+      pod
+
+    print_message "Migrating vault-${i}"
+    kubectl exec -it "vault-${i}" -- \
+      sh -c "vault operator unseal -migrate \"${OLD_VAULT_UNSEAL_KEY_1}\""
+    kubectl exec -it "vault-${i}" -- \
+      sh -c "vault operator unseal -migrate \"${OLD_VAULT_UNSEAL_KEY_2}\""
+    kubectl exec -it "vault-${i}" -- \
+      sh -c "vault operator unseal -migrate \"${OLD_VAULT_UNSEAL_KEY_3}\""
+  done
+  ```
+
+### 9. Step down leader pod
+
+Once all follower pods have been migrated, the leader must step down so it can
+restart and apply the new seal configuration.
+
+- Step down the leader pod
+
+  ```bash
+  kubectl exec -it vault-0 -- vault login "${OLD_VAULT_MIGRATION_TOKEN}"
+  kubectl exec -it vault-0 -- vault operator step-down
+  ```
+
+### 10. Remove old seal block from the config
+
+Now that all pods use the new seal configuration, we can safely remove the old
+seal block.
+
+- Edit the config map
+
+  ```bash
+  kubectl get cm vault-config -o yaml > vault-config.yaml
+
+  ed '+/disabled \+= true/' vault-config.yaml <<EOF
+  -
+  .;/}/d
+  w
+  q
+  EOF
+
+  kubectl apply -f vault-config.yaml
+  ```
+
+### 11. Restart all replicas
+
+- Restart all Vault pods to ensure they all use the final configuration.
+  ```bash
+  kubectl delete pod --selector=app.kubernetes.io/name=vault
+  ```
+
+### 12. Create a proxy service from the old Vault to the new Vault
+
+To allow clients using the old Vault domain to reach the new Vault instance, we
+create a proxy `Service` and `EndpointSlice` in the old cluster that forwards
+traffic to the new Vault’s ingress.
```

Comment:
We should probably mention that `ExternalName` in ingress-nginx is limited to licensed uses of nginx, hence we need to use an IP address instead of a DNS name.

Reply:
Agreed. Done.